app/src/main/java/org/olpc/xorobothash/MainActivity.java

```diff
@@ -0,0 +1,164 @@
+        if (id == R.id.action_settings) {
+            return true;
+        }
+
+        return super.onOptionsItemSelected(item);
+    }
+
+
+    private void getRobot() {
+        Log.i (tag, "In getRobot()");
+        //Check if the robot image file already exists
+        File file = new File(robotImagePath);
+        // Check if we have already downloaded the XO's robot image
+        if(file.exists()) {
+            Log.i (tag, "Robohash image file exists.");
+            // Show the already downloaded robohash image
+            showRobot(robotImagePath);
+            return;
+        }
+
+        Log.i (tag, "File does not exist.");
+        // We haven't already downloaded the image - check network connectivity
+        ConnectivityManager connMgr = (ConnectivityManager)
+                getSystemService(Context.CONNECTIVITY_SERVICE);
+        NetworkInfo networkInfo = connMgr.getActiveNetworkInfo();
+        if (networkInfo != null && networkInfo.isConnected()) {
+            Log.i (tag, "Network is connected.");
+            // Get the device's serial and concatenate into a RoboHash URL
+            String serialNumber = android.os.Build.SERIAL;  // if SERIAL doesn't work, try Settings.Secure.ANDROID_ID; ?
+            String roboHashUrl = getString(R.string.robohash_url) + serialNumber + getString(R.string.robohash_file_extension);
+            // Execute the async download of the image
+            new DownloadRobotImage().execute(roboHashUrl);
+        } else {
+            Log.e (tag, "Network is NOT connected.");
+            // Hide the "Your robot is" message
+            TextView robotText = (TextView) findViewById(R.id.robot_text_message);
+            robotText.setVisibility(View.GONE);
+            // Show the no network connection message
+            TextView errorMessage = (TextView) findViewById(R.id.error_message);
+            errorMessage.setText(getString(R.string.no_network_connection));
+            return;
+        }
+
+    }
+
+    private void showRobot(String imagePath) {
+        Log.i (tag, "In showRobot()");
+        if(imagePath!=null) {
+            Bitmap bitmap = BitmapFactory.decodeFile(imagePath);
+            robotImageView.setImageBitmap(bitmap);
+        }
+    }
+
+    private class DownloadRobotImage extends AsyncTask<String, Void, String> {
+
+        @Override
+        protected void onPreExecute() {
+            // TODO Display a spinner or loading message
+        }
+
+        @Override
+        protected String doInBackground(String... params) {
+            Log.i (tag, "In DownloadRobotImage() doInBackground()");
+
+            HttpURLConnection connection = null;
+
+            try {
+                URL url = new URL(params[0]);
+                connection = (HttpURLConnection) url.openConnection();
+
+                InputStream in = new BufferedInputStream(connection.getInputStream());
+
+                OutputStream out = new FileOutputStream(robotImagePath);
```

Comment:
Rather than access the `robotImagePath` in scope here, pass it as a parameter like you did with the URL.